Core/src/Geometry/GridPortalLinkMerging.cpp

```diff
@@ -0,0 +1,569 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Definitions/Tolerance.hpp"
+#include "Acts/Geometry/CompositePortalLink.hpp"
+#include "Acts/Geometry/GridPortalLink.hpp"
+#include "Acts/Surfaces/CylinderSurface.hpp"
+#include "Acts/Surfaces/DiscSurface.hpp"
+#include "Acts/Surfaces/PlaneSurface.hpp"
+#include "Acts/Surfaces/RadialBounds.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <limits>
+#include <memory>
+#include <stdexcept>
+
+namespace Acts {
+
+namespace {
+
+template <typename... Args>
+std::unique_ptr<GridPortalLink> makeGrid(
+    const std::shared_ptr<RegularSurface>& surface, BinningValue direction,
+    const Logger& logger, std::tuple<Args...> args, const IAxis* otherAxis,
+    bool prepend) {
+  // @TODO: PlaneSurface support
+
+  ACTS_VERBOSE("Make resulting merged grid");
+
+  // This is to make it possible to construct Axis with the tuple arguments
+  auto axisFactory = [](auto&&... axisArgs) {
+    return Axis{std::forward<decltype(axisArgs)>(axisArgs)...};
+  };
+
+  // Avoid copy-pasting identical code twice below
+  auto makeGrid = [&](auto boundaryType) -> std::unique_ptr<GridPortalLink> {
+    auto axisArgs = std::tuple_cat(std::tuple{boundaryType}, std::move(args));
+    auto merged = std::apply(axisFactory, std::move(axisArgs));
+
+    if (otherAxis == nullptr) {
+      // No other axis
+      ACTS_VERBOSE("    ~> merged axis: " << merged);
+      return GridPortalLink::make(surface, direction, std::move(merged));
+    } else {
+      return otherAxis->visit(
+          [&](const auto& axis) -> std::unique_ptr<GridPortalLink> {
+            if (prepend) {
+              ACTS_VERBOSE("    ~> other axis (prepend): " << axis);
+              ACTS_VERBOSE("    ~> merged axis: " << merged);
+              return GridPortalLink::make(surface, axis, std::move(merged));
+            } else {
+              ACTS_VERBOSE("    ~> merged axis: " << merged);
+              ACTS_VERBOSE("    ~> other axis (append): " << axis);
+              return GridPortalLink::make(surface, std::move(merged), axis);
+            }
+          });
+    }
+  };
+
+  // Check if we're in the cylinder or disc case, and the resulting bounds wrap
+  // around and should have closed binning
+  if (direction == BinningValue::binPhi || direction == BinningValue::binRPhi) {
+    if (const auto* cylinder =
+            dynamic_cast<const CylinderSurface*>(surface.get());
+        cylinder != nullptr) {
+      if (cylinder->bounds().coversFullAzimuth()) {
+        return makeGrid(AxisClosed);
+      }
+    } else if (const auto* disc =
+                   dynamic_cast<const DiscSurface*>(surface.get());
+               disc != nullptr) {
+      const auto& radialBounds =
+          dynamic_cast<const RadialBounds&>(disc->bounds());
+      if (radialBounds.coversFullAzimuth()) {
+        return makeGrid(AxisClosed);
+      }
+    }
+  }
+
+  return makeGrid(AxisBound);
```

Comment:
That's a value of a helper type that I untroduced back in https://github.com/acts-project/acts/pull/3257

https://github.com/acts-project/acts/blob/2cfa5f74720d53eb4f70246049e38231767c4c42/Core/include/Acts/Utilities/AxisFwd.hpp#L36-L38

I use it here because I don't have to specify the template parameter this way.